src/reference/asciidoc/gateway.adoc

```diff
@@ -468,9 +470,92 @@ public interface NoExecGateway {
 
 IMPORTANT: If the return type is a specific concrete `Future` implementation or some other subinterface that is not supported by the configured executor, the flow will run on the caller's thread and the flow must return the required type in the reply message payload.
 
-*Asynchronous Gateway and Reactor Promise*
+[[gw-completable-future]]
+===== CompletableFuture
+
+Starting with _version 4.2_, gateway methods can now return `CompletableFuture<?>`.
+There are several modes of operation when returning this type:
+
+When an async executor is provided *and* the return type is exactly `CompletableFuture` (not a subclass), the framework
+will run the task on the executor and immediately return a `CompletableFuture` to the caller.
+`CompletableFuture.supplyAsync(Supplier<U> supplier, Executor executor)` is used to create the future.
+
+When the async executor is explicitly set to `null` and the return type is `CompletableFuture` *or* the return type
+is a subclass of `CompletableFuture`, the flow is invoked on the caller's thread.
+In this scenario, it is expected that the downstream flow will return a `CompletableFuture` of the appropriate type.
+
+*Usage Scenarios*
+
+[source, java]
+----
+
+CompletableFuture<Invoice> order(Order order);
+----
+
+[source, xml]
+----
+
+<int:gateway service-interface="foo.Service" default-request-channel="orders" />
+----
+
+In this scenario, the caller thread returns immediately with a `CompletableFuture<Invoice>` which will be completed
+when the downstream flow replies to the gateway (with an `Invoice` object).
+
+[source, java]
+----
+
+CompletableFuture<Invoice> order(Order order);
+----
+
+[source, xml]
+----
+
+<int:gateway service-interface="foo.Service" default-request-channel="orders"
+    async-executor="" />
+----
+
+In this scenario, the caller thread will return with a CompletableFuture<Invoice> when the downstream flow provides
+it as the payload of the reply to the gateway.
+Some other process must complete the future when the invoice is ready.
+
+[source, java]
+----
+
+MyCompletableFuture<Invoice> order(Order order);
+----
+
+[source, xml]
+----
+
+<int:gateway service-interface="foo.Service" default-request-channel="orders" />
+----
+
+In this scenario, the caller thread will return with a CompletableFuture<Invoice> when the downstream flow provides
+it as the payload of the reply to the gateway.
+Some other process must complete the future when the invoice is ready.
+If `DEBUG` logging is enabled, a log is emitted indicating that the async executor cannot be used for this scenario.
+
+
+`CompletionFuture` s can be used to perform additional manipulation on the reply, such as:
```

Comment:
Is this intentional word?
" `CompletableFuture` s " ?
Let me know and I fix it and merge.